File: c/perfect-numbers/perfect_numbers.c
#include "perfect_numbers.h"

kind classify_number(int64_t number) {
    kind result = ERROR;

    if (number <= 0) {
        return result;
    }

    factors_list_t numbers = factors(number);
    int64_t aliquotSum = sum(numbers);

    if (aliquotSum > number) {
        result = ABUNDANT_NUMBER;
    } else if (aliquotSum < number) {
        result = DEFICIENT_NUMBER;
    } else {
        result = PERFECT_NUMBER;
    }

    return result;
}

int64_t sum(const factors_list_t numbers) {
    int64_t total = 0;

    if (numbers.size <= 0) {
        return total;
    }

    for (size_t i = 0; i < numbers.size; i++) {
        total += numbers.factors[i];
    }

    return total;
}

factors_list_t factors(int64_t number) {
    factors_list_t numbers;
    numbers.size = 0;

    if (number <= 0) {
        return numbers;
    }

    for (int64_t factor = 1; factor < number; factor++) {
        if (number % factor == 0) {
            numbers.factors[numbers.size++] = factor;
        }
    }

    return numbers;
}

File: ruby/perfect-numbers/perfect_numbers.rb
# frozen_string_literal: false

# https://exercism.org/tracks/ruby/exercises/perfect-numbers
# Perfect Numbers exercise
module PerfectNumber
  Struct.new('Kind', :aliquot_sum, :number) do
    def deficient
      -1
    end

    def perfect
      0
    end

    def abundant
      1
    end

    def kind
      if aliquot_sum > number
        abundant
      elsif aliquot_sum < number
        deficient
      else
        perfect
      end
    end

    def to_s
      kinds = {
        deficient => 'deficient',
        perfect => 'perfect',
        abundant => 'abundant'
      }

      kinds[kind]
    end

    def to_i
      kind
    end

    private :deficient, :perfect, :abundant, :kind
  end

  def self.classify(number)
    raise 'perfect numbers are postive' unless number.positive?

    aliquot_sum = (1...number).reduce do |sum, factor|
      sum + (number.modulo(factor).zero? ? factor : 0)
    end

    Struct::Kind.new(aliquot_sum, number).to_s
  end
end
